### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by fredguth** [jacktrip-webrtc/jacktrip-webrtc#30] [Ideas]
I wonder what are the next steps for this project.  I think it would be interesting to brainstorm some ideas and build a collective Roadmap (we could even use Github Projects for this).

The current implementation shows the limitations of using WebRTC for Network Musical Performances (NMP) today. We are talking about 55 ms only due to WebRTC audio capturing and playing processing (see discussion in #28).  

There is a long discussion #9 about using jack, which would not prevent the 55ms latency (for usage in the web with webrtc), but can make a lot of difference in the networking latency that depends on distance but is probably the biggest problem. The idea of doing the processing of the worklets in wasm (#28)  does not seem to help much as the biggest benefit was the separation of this thread from the js main event loop which was already done with web workers.

In #9 there was also an interesting idea from @noahhorn: "... use lower-level computing processes to deal with the audio, i.e. bypass the web browser for this, which is what software like JackTrip, Jamulus, etc. do. But... keep the web browser involved for ease of use...  namely, ... a Node app that would be deployed with Electron, which could also include the full binaries for JackTrip, Jack, etc. This app displayed a smooth interface similar to jacktrip-webrtc, but under the hood it was parsing JackTrip's text output and trying to make connections and handle NAT traversal, etc. It's basically a wrapper for JackTrip that uses Electron & Node to allow for full HTML5/CSS/JS capabilities for the GUI, and I actually tied it into a PHP/MySQL/Apache web server that I have, in order to handle user accounts. Matteo's code for traversing NAT, and handling peer-to-peer connections, is all very valuable for this, along with the concept of rooms, etc."

It seems interesting. I would certainly contribute to that.    

I have seen other projects that claim to get very low latencies (like FarPlay), but I don't see a third-party review of the latencies.  If someone has experience using them, I would be glad to learn about your experiences.

Also, I would love to find some research on the impact on sound quality *perception* due to different lossy compression methods (using a codec like Opus, or simply reducing the sample rate).  

What are your thoughts on the subject? What is the path moving forward?  Where lay your interests?

**Comment by matteosacchetto** [comment]
Hi Fred,
thank you for the interest shown towards this project.

As it is for now the "limitations" in terms of the lowest possible latency seem to be related to the acquisition/playback steps. While it is true that WebRTC is involved in this process (only acquisition though), I think that the root cause why we are not able to reach the lowest possible latency, seem to be less related to WebRTC itself, but mostly due to the fact that a web browser, being an application which is not developed for low latency audio processing, has not an easy way to access the low latency audio API provided by the OS, and so, like any other audio application, it runs on top of the system-wide audio API provided by the OS, which is not targeted for low latency in general
So all comes down to how good is the OS in providing low latency for the application running on top of that API.

While it is an idea to use try to use Jack for reducing that latency, we did not follow that route mainly because is not a solution which is easy to use for the user. The idea behind the project was to achieve an ease of use which is similar to the one of web applications used for video conferencing, so not requiring the use to install anything on their system, and not requiring them to tinker with the configuration of their system, those are the reasons which lead us to not follow the path of requiring the use to install something on their system, which sometimes is not as straightforward to do.

As it is for now we are trying to improve scalability of the application (this is something which we are already working on), which is a problem in the v1 of the app, and we are trying to have a better understanding of all the details about the latencies we are experiencing and the different contributions which add up to the final latency we measured.

Matteo
